latex/DPIA/primitives.tex

```diff
@@ -1,96 +1,66 @@
             \to \tyexp{\dttwo}{\writeacc}
             \to \tyexp{\lArray{\n}{\dtone}}{\readacc}
             \to \tyexp{\dttwo}{\readacc}\\
 
           \prim{reduceSeqUnroll}&:&(\n : \natkind) \to (\dtone~\dttwo : \datatype) \to\\
-            &&\qquad (\tyexp{\dtone}{\readacc} \to \tyexp{\dttwo}{\readacc} \to \tyexp{\dttwo}{\writeacc})
+            &&\qquad (\tyexp{\dttwo}{\readacc} \to \tyexp{\dtone}{\readacc} \to \tyexp{\dttwo}{\writeacc})
             \to \tyexp{\dttwo}{\writeacc}
             \to \tyexp{\lArray{\n}{\dtone}}{\readacc}
             \to \tyexp{\dttwo}{\readacc}\\
+          \\[-.75em]
+
+          \prim{idx}&:& (\n: \natkind) \to (\dt: \datatype)
+            \to \tyexp{\tyidx[\n]}{\readacc}
+            \to \tyexp{\lArray{\n}{\dt}}{\readacc}
+            \to \tyexp{\dt}{\readacc}\\
+          \\[-.75em]
 
           \prim{scanSeq}&:& (\n : \natkind) \to (\dtone~\dttwo : \datatype)
             \to (\tyexp{\dtone}{\readacc} \to \tyexp{\dttwo}{\readacc} \to \tyexp{\dttwo}{\writeacc}) \to\\
             &&\qquad \tyexp{\dttwo}{\writeacc}
             \to \tyexp{\lArray{\n}{\dtone}}{\readacc}
             \to \tyexp{\lArray{\n}{\dttwo}}{\readacc}\\
-
-          \prim{iterate}&:& (\n~\m~\nat{k}: \natkind) \to (\dt: \datatype) \to\\
-            &&\qquad ((\nat{l}: \natkind) \to \tyexp{\lArray{\nat{l}\n}{\dt}}{\readacc} \to \tyexp{\lArray{\nat{l}}{\dt}}{\readacc})
-            \to \tyexp{\lArray{\m\cdot\n^{\nat{k}}}{\dt}}{\readacc}
-            \to \tyexp{\lArray{\m}{\dt}}{\readacc}\\
-
-          \medskip\\
-
-          \prim{split}&:& (\n~\m : \natkind)\to (\wvar : \access) \to (\dt : \datatype)
-            \to \tyexp{\lArray{\n\m}{\dt}}{\wvar}
-            \to \tyexp{\lArray{\m}{\lArray{\n}{\dt}}}{\wvar}\\
-
-          \prim{join}&:& (\n~\m: \natkind) \to (\wvar: \access) \to (\dt: \datatype)
-            \to \tyexp{\lArray{\n}{\lArray{\m}{\dt}}}{\wvar}
-            \to \tyexp{\lArray{\n\m}{\dt}}{\wvar}\\
-          \\[-.75em]
-
-          \prim{slide}&:& (\n~\nat{sz}~\nat{sp} : \natkind) \to (\dt : \datatype)
-            \to \tyexp{\lArray{\nat{(sp \cdot n + sz - sp)}}{\dt}}{\readacc}
-            \to \tyexp{\lArray{\n}{\lArray{\nat{sz}}{\dt}}}{\readacc}\\
-
-          \prim{slideSeq}&:& (\n~\nat{sz}~\nat{sp} : \natkind) \to (\dt : \datatype)
-            \to \tyexp{\lArray{\nat{(sp \cdot n + sz - sp)}}{\dt}}{\readacc}
-            \to \tyexp{\lArray{\n}{\lArray{\nat{sz}}{\dt}}}{\readacc}\\
-          \\[-.75em]
-
-          \prim{reorder}&:& (\n: \natkind) \to (\dt: \datatype)
-            \to (\tyexp{\tyidx[\n]}{\readacc} \to \tyexp{\tyidx[\n]}{\readacc}) \to\\
-            &&\qquad (\tyexp{\tyidx[\n]}{\readacc} \to \tyexp{\tyidx[\n]}{\readacc})
-            \to \tyexp{\lArray{\n}{\dt}}{\readacc}
-            \to \tyexp{\lArray{\n}{\dt}}{\readacc}\\
-
-          \prim{cycle}&:& (\n~\m: \natkind) \to (\dt: \datatype)
-            \to \tyexp{\lArray{\m}{\dt}}{\readacc}
-            \to \tyexp{\lArray{\n}{\dt}}{\readacc}\\
```

Comment:
is there a logic behind this reordering?

Reply:
yes. When you look at the document you see that `write` primitives (those which return a writeable expression), read and primitives polymorphic over the acccess type are grouped together.